### PR TITLE
Pass also Session

### PR DESCRIPTION
**Pull request by dereuromark**
Tackles the issue of callback now not being able to access controller/component anymore (coming from Hybrid)
https://github.com/ADmad/cakephp-social-auth/issues/70

This way one could manually write the necessary Flash message into the session before redirect
And we wouldn't have to mess with the query string ? success topic.

### CODE REVIEW

**Comment by dereuromark** [comment]
This is what we could also documented and works:

```php
    /**
     * @param \Cake\Http\MiddlewareQueue $middlewareQueue The middleware queue to setup.
     *
     * @return \Cake\Http\MiddlewareQueue The updated middleware queue.
     */
    public function middleware(MiddlewareQueue $middlewareQueue): MiddlewareQueue
    {
        EventManager::instance()->on('SocialAuth.afterIdentify', [$this, 'afterIdentify']);

        ...
```
Further down in the Application.php:
```php
    /**
     * @param \Cake\Event\EventInterface $event
     * @param \App\Model\Entity\User $user
     * @param \Cake\Http\Session $session
     *
     * @return null
     */
    public function afterIdentify(EventInterface $event, User $user, Session $session)
    {
        $messages = (array)$session->read('Flash.flash');
        $messages[] = [
            'message' => __('You are now logged in'),
            'key' => 'flash',
            'element' => 'flash/success',
            'params' => [],
        ];
        $session->write('Flash.flash', $messages);

        return null;
    }
```

**Comment by dereuromark** [comment]
Are you able to tag a beta or RC on this one then please?
We would like to start deploying at work the next days here.

File: src/Middleware/SocialAuthMiddleware.php [hunk 1]
```diff
@@ -225,7 +225,8 @@ protected function _handleCallbackAction(ServerRequest $request): Response
 
         $user->unset($config['fields']['password']);
 
-        $event = $this->dispatchEvent(self::EVENT_AFTER_IDENTIFY, ['user' => $user]);
+        $data = ['user' => $user, 'session' => $request->getSession()];
```
Comment (ADmad) [inline src/Middleware/SocialAuthMiddleware.php:228]: I am not happy with passing session instance as event data. As I commented on the other issue I would rather use query string.

> I wonder if that could also conflict with other query strings and stuff.

The AuthComponent and new Auth plugin uses `redirect` query string var to retrain the URL to redirect to after successful login (change done by yourself). No one has complained about that causing any issues, so I don't see this being any different.

> Or we would need to make the query string a bit more unique and less collision likely.

Yes, I am open to suggestions for what var name to use.

Comment (dereuromark) [reply src/Middleware/SocialAuthMiddleware.php:228]: redirect is a global default expectation IMO.
that one not so much (yet)

How about both possible?
I would rather use the internal session handling than the query string for that case, and it is as "safe" as the other operations using direct session access IMO, e.g. getUser() event.
We can however document primarily the query string one if we want to.

Comment (dereuromark) [reply src/Middleware/SocialAuthMiddleware.php:228]: Also redirect is a different topic altogether (I am well aware of that being my thing^^).
It is added to the login action only and also removed again when redirecting to the actual controller actions across the whole app and plugins.

The query string would directly be visible and can have annoying side effects
E.g. reload would give the same message again if you don't redirect and remove that one.

So I really would advise against using query string here.

Comment (ADmad) [reply src/Middleware/SocialAuthMiddleware.php:228]: Okay, but I have something a bit different in mind to achieve the same. I will make a PR soon.

**Pull request closed without merging**